### PR TITLE
Readersearch: add a button to call a new search input dialog

### PR DESCRIPTION
**Pull request by hius07**
Allows to call a new search from the continuous search button dialog.

<kbd></kbd>

Icons in the buttons are supported in button.lua, but were disabled in buttontable.lua.
Icons in the buttons are smaller (than icons in IconButtons) to correspond with text buttons.
This will allow to use standard KOReader icons in buttons without resizing.
Closes https://github.com/koreader/koreader/issues/8188.

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/8190)

### CODE REVIEW

**Comment by hius07** [comment]
Without `nextTick` I had (with flash_ui enabled)

<kbd></kbd>

File: frontend/ui/widget/button.lua [hunk 1]
```diff
@@ -36,8 +36,8 @@ local Button = InputContainer:new{
     text = nil, -- mandatory
     text_func = nil,
     icon = nil,
-    icon_width = Screen:scaleBySize(DGENERIC_ICON_SIZE), -- our icons are square
-    icon_height = Screen:scaleBySize(DGENERIC_ICON_SIZE),
+    icon_width = Screen:scaleBySize(DGENERIC_ICON_SIZE * 0.8), -- our icons are square
+    icon_height = Screen:scaleBySize(DGENERIC_ICON_SIZE * 0.8),
```
Comment (poire-z) [inline frontend/ui/widget/button.lua:40]: Might be worth a comment, if these 0.8 are really to match the size of "cfont" size 20 - and that all is worse if you change any of these 2 without tuning the other. Can you test that this is really the case ?

File: frontend/apps/reader/modules/readersearch.lua [hunk 2]
```diff
@@ -303,6 +303,19 @@ function ReaderSearch:onShowSearchDialog(text, direction, regex, case_insensitiv
...
+                {
+                    icon = "appbar.search",
+                    vsync = true,
+                    callback = function()
+                        self.ui.highlight:clear()
+                        UIManager:setDirty(self.dialog, "ui")
+                        UIManager:nextTick(function()
+                            UIManager:close(self.search_dialog)
+                        end)
+                        self.last_search_text = text
```
Comment (zwim) [inline frontend/apps/reader/modules/readersearch.lua:315]: Is line 315 really necessary?

Comment (hius07) [reply frontend/apps/reader/modules/readersearch.lua:315]: Yes, when calling Search from the highlight popup menu.

**Comment by zwim** [comment]
> Without `nextTick` I had (with flash_ui enabled)
> 

This seems to be an artifact of the sandhour.
I also had/have sometimes similar artifacts in the emulator, but not on the device. Do you want to check how your PR behaves on a device with an without the nexttick?

File: frontend/ui/widget/button.lua (hunk 1 above)
Comment (hius07) [reply frontend/ui/widget/button.lua:40]: Yes, the 0.8 factor corresponds only to cfont-20
https://github.com/koreader/koreader/blob/0283c44ff7b0904b02f29ab74574930c76f9d1a4/frontend/ui/widget/buttontable.lua#L25-L26
So changing buttons font face/size will cause the icon to look smaller/bigger than the text in other buttons.
Do you mean just adding a comment or to recalculate the factor depending on the font size of other text buttons in buttontable?

**Comment by hius07** [comment]
I always test on my Kindle Voyage only (have no emulator).

File: frontend/ui/widget/button.lua (hunk 1 above)
Comment (poire-z) [reply frontend/ui/widget/button.lua:40]: No need to recalculate if we don't use a font other than the default anywhere (if someone needs another font size - and he shouldn't :) - he'll do the maths himself :)
Just a comment, may be move the text related settings together above the icons related settings, like this:
```lua
local Button = InputContainer:new{
    text = nil, -- mandatory (unless icon is provided)
    text_func = nil,
    text_font_face = "cfont",
    text_font_size = 20,
    text_font_bold = true,
    avoid_text_truncation = true,
    icon = nil,
    icon_width = Screen:scaleBySize(DGENERIC_ICON_SIZE * 0.8), -- our icons are square
    icon_height = Screen:scaleBySize(DGENERIC_ICON_SIZE * 0.8),
        -- These *0.8 make our icons nicely match the size of the default font
        -- above ("cfont", 20), when mixing both text and icon buttons.
    icon_rotation_angle = 0,
```

**Comment by NiLuJe** [comment]
Try requesting a refresh via the close instead of relying on nextTick (c.f., docs, as I'm on mobile right now, and there are very few examples of it in the code ;)).

**Comment by hius07** [comment]
Fixed with removing `vsync = true`.

**Comment by NiLuJe** [comment]
That won't do, it's necessary ;).

**Comment by NiLuJe** [comment]
@hius07: I'm very sorry, I had misconstrued what this was supposed to be doing... Disabling `vsync` was indeed the right thing to do, as this closes the search bar ;).

So, simplified to avoid code duplication, this leads us to:

```diff
diff --git a/frontend/apps/reader/modules/readersearch.lua b/frontend/apps/reader/modules/readersearch.lua
index 83ab8c3b7..156bc76fe 100644
--- a/frontend/apps/reader/modules/readersearch.lua
+++ b/frontend/apps/reader/modules/readersearch.lua
@@ -305,13 +305,8 @@ function ReaderSearch:onShowSearchDialog(text, direction, regex, case_insensitiv
                 },
                 {
                     icon = "appbar.search",
-                    vsync = true,
                     callback = function()
-                        self.ui.highlight:clear()
-                        UIManager:setDirty(self.dialog, "ui")
-                        UIManager:nextTick(function()
-                            UIManager:close(self.search_dialog)
-                        end)
+                        self.search_dialog:onClose()
                         self.last_search_text = text
                         self:onShowFulltextSearchInput()
                     end,
```

**Comment by hius07** [comment]
Thank you, we also must clear old highlighting before calling new search input dialog.

**Comment by NiLuJe** [comment]
That's handled in `onClose` via the `tap_close_callback` (thgat's what I meant by code duplication) ;).

**Comment by hius07** [comment]
Oh, I see, thanks!

**Comment by hius07** [comment]
So, ready.

**Review by NiLuJe — APPROVED** [review]

**Comment by hius07** [comment]
I'm very sorry for a mistake, reducing icon size has affected page turn chevrons in file browser and sorting widget:
https://github.com/koreader/koreader/blob/e16d36d5da658c18871789bbb300f95d835c7bb1/frontend/ui/widget/menu.lua#L769-L774
I will fix it soonest.

**Pull request merged**